File: app/src/main/java/seigneur/gauvain/mycourt/ui/shotEdition/EditShotActivity.java
package seigneur.gauvain.mycourt.ui.shotEdition;

import android.Manifest;
import android.app.Application;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.ActivityCompat;
import android.os.Bundle;
import android.support.v4.content.FileProvider;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.Html;
import android.text.TextWatcher;
import android.widget.Button;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.yalantis.ucrop.UCrop;

import java.io.File;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import dagger.android.AndroidInjection;
import seigneur.gauvain.mycourt.R;
import seigneur.gauvain.mycourt.data.model.Shot;
import seigneur.gauvain.mycourt.data.model.ShotDraft;
import seigneur.gauvain.mycourt.ui.base.BaseActivity;
import seigneur.gauvain.mycourt.utils.Constants;
import seigneur.gauvain.mycourt.utils.image.ImagePicker;
import seigneur.gauvain.mycourt.ui.widget.FourThreeImageView;
import seigneur.gauvain.mycourt.utils.image.ImageUtils;
import seigneur.gauvain.mycourt.utils.MyTextUtils;
import timber.log.Timber;

public class EditShotActivity extends BaseActivity {

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    @Inject
    Application mApplication;

    private ShotEditionViewModel mShotEditionViewModel;

    @BindView(R.id.toolbar)
    Toolbar mToolbar;

    @BindView(R.id.main)
    CoordinatorLayout layoutMain;

    @BindView(R.id.shot_title_edt)
    TextInputEditText mShotTitleEditor;

    @BindView(R.id.shot_tag_edt)
    TextInputEditText mTagEditor;

    @BindView(R.id.shot_description_edt)
    TextInputEditText mShotDescriptionEditor;

    @BindView(R.id.cropped_img_preview)
    FourThreeImageView croppedImagePreview;

    @BindView(R.id.btn_store)
    Button storeBtn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_shot);
        ButterKnife.bind(this);
        //provide Dependencies
        AndroidInjection.inject(this);
        //Listen EditText
        mShotTitleEditor.addTextChangedListener(titleWatcher);
        mTagEditor.addTextChangedListener(tagWtacher);
        mShotDescriptionEditor.addTextChangedListener(descriptionWatcher);
        //Provide ViewModel
        mShotEditionViewModel = ViewModelProviders.of(this, viewModelFactory).get(ShotEditionViewModel.class);
        mShotEditionViewModel.init();
        //Subscribe to ViewModel data and event
        subscribeToLiveData(mShotEditionViewModel);
        subscribeToSingleEvent(mShotEditionViewModel);

    }

    @OnClick(R.id.btn_store)
    public void store() {
        mShotEditionViewModel.onStoreDraftClicked();
    }

    @OnClick(R.id.btn_publish)
    public void publish() {
        mShotEditionViewModel.onPublishClicked();
    }

    /*
     *********************************************************************************************
     * EVENT WHICH VIEW WILL SUBSCRIBE
     *********************************************************************************************/
    private void subscribeToLiveData(ShotEditionViewModel viewModel) {
        viewModel.getCroppedImageUri().observe(this, this::displayShotImagePreview);

        viewModel.getTitle().observe(this, title -> {
            Timber.d("title change:" +title);
        });

        viewModel.getDescription().observe(this,  desc -> {
            Timber.d("desc change:" +desc);
        });

        viewModel.getTags().observe(this, tags -> {
            Timber.d("tags change:" +tags);
        });
    }

    private void subscribeToSingleEvent(ShotEditionViewModel viewModel) {

        viewModel.getSetUpUiCmd().observe(
                this,
                call -> setUpEditionUI(mShotEditionViewModel.getObjectSource())
        );

        viewModel.getPickShotCommand().observe(this, call -> openImagePicker());

        viewModel.getCropImageCmd().observe(this,
                call -> goToUCropActivity(
                        mShotEditionViewModel.getImagePickedFormat(),
                        mShotEditionViewModel.getImagePickedUriSource(),
                        mShotEditionViewModel.getImagePickedFileName(),
                        mShotEditionViewModel.getImageSize()));

        viewModel.getPickCropImgErrorCmd().observe(this,
                errorCode -> Toast.makeText(this, "oops :"+errorCode , Toast.LENGTH_SHORT).show());

        viewModel.getCheckPerm().observe(this,
                call -> checkPermissionExtStorage());

        viewModel.getRequestPermCmd().observe(this,
                call -> requestPermission());

    }

    /*
     *********************************************************************************************
     * INNER
     *********************************************************************************************/
    @OnClick(R.id.cropped_img_preview)
    public void shotPreviewClick() {
        mShotEditionViewModel.onImagePreviewClicked();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_OK) {
            if (requestCode== Constants.PICK_IMAGE_REQUEST) {
                mShotEditionViewModel.setImagePickedUriSource(ImagePicker.getImageUriFromResult(this, resultCode, data));
                mShotEditionViewModel.setImagePickedFileName(ImagePicker.getPickedImageName(this, mShotEditionViewModel.getImagePickedUriSource()));
                mShotEditionViewModel.setImagePickedFormat(ImageUtils.getImageExtension(this, mShotEditionViewModel.getImagePickedUriSource()));
                mShotEditionViewModel.setImageSize(ImageUtils.imagePickedWidthHeight(this, mShotEditionViewModel.getImagePickedUriSource(), 0));

                mShotEditionViewModel.onImagePicked();
            }
            else
            if (requestCode == UCrop.REQUEST_CROP)
                mShotEditionViewModel.onImageCropped(UCrop.getOutput(data));
        } else {
            if (resultCode == UCrop.RESULT_ERROR)
                mShotEditionViewModel.onPickcropError(0);
            else
                mShotEditionViewModel.onPickcropError(1);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    /**
     * Callback received when a permissions request has been completed.
     * @param requestCode
     * @param permissions
     * @param grantResults
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case Constants.REQUEST_STORAGE_WRITE_ACCESS_PERMISSION:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    mShotEditionViewModel.onPermGranted();
                } else {
                    mShotEditionViewModel.requestPerm();
                }
                break;
            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    public void checkPermissionExtStorage() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            //todo single live event
            mShotEditionViewModel.requestPerm();
        } else {
            mShotEditionViewModel.onPermGranted();
        }
    }

    public void requestPermission() {
        requestPermission(
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                getString(R.string.permission_write_storage_rationale),
                Constants.REQUEST_STORAGE_WRITE_ACCESS_PERMISSION);
    }

    public void notifyPostSaved() {
        Toast.makeText(this, "ShotDraft saved in memory: ", Toast.LENGTH_SHORT).show();
        finishAfterTransition();
    }

    public void openImagePicker() {
        ImagePicker.pickImage(this, Constants.PICK_IMAGE_REQUEST);
    }

    public void goToUCropActivity(String imagePickedformat,
                                  Uri imagePickedUriSource,
                                  String imagePickedFileName,
                                  int[] imageSize) {
        ImageUtils.goToUCropActivity(imagePickedformat,
                imagePickedUriSource,
                Uri.fromFile(new File(getCacheDir(), imagePickedFileName)),this,
                imageSize);
    }

    public void showImageNotUpdatable() {
        Toast.makeText(EditShotActivity.this, "Shot can't be changed in edition mode", Toast.LENGTH_SHORT).show();
    }

    public void openConfirmMenu() {
    }

    public void stopActivity() {
        finish();
    }

    /*
     *********************************************************************************************
     * UI - MANAGE EDITION MODE
     *********************************************************************************************/
    private void setUpEditionUI(Object object) {
        if (object!=null) {
            if (EditUtils.getDraftType(object)==Constants.EDIT_MODE_UPDATE_SHOT)
                mToolbar.setTitle("Edit a shot");
            else
                mToolbar.setTitle("Create a shot");
            //notify viewModel so we Ui display image
            Uri imageUri = EditUtils.getImageUrl(this, object);
            mShotEditionViewModel.onImageCropped(imageUri);
            //notify view model about format of image
            mShotEditionViewModel.setImagePickedFormat(EditUtils.getImageFormat(object));
            //set up String in Edit Text
            mShotTitleEditor.setText(EditUtils.getTitle(object));
            String description = EditUtils.getDescription(object);
            if (description!=null)
                mShotDescriptionEditor.setText(MyTextUtils.noTrailingwhiteLines(description));

            mTagEditor.setText(EditUtils.getTagList(object));
        } else {
            mToolbar.setTitle("Create a shot");
            //set an image but doesn't notify viewmodel
            croppedImagePreview.setImageResource(R.drawable.add_image_illustration);
        }

    }

    private void displayShotImagePreview(Uri uriImageCropped) {
        if (uriImageCropped!=null) {
            Glide.with(mApplication)
                    .asBitmap()
                    .load(uriImageCropped)
                    .apply(new RequestOptions()
                                    .error(R.drawable.ic_my_shot_black_24dp)
                            // .diskCacheStrategy(DiskCacheStrategy.ALL)
                    )
                    .listener(new RequestListener<Bitmap>() {
                        @Override
                        public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Bitmap> target, boolean isFirstResource) {
                            Toast.makeText(EditShotActivity.this, "error loading image", Toast.LENGTH_SHORT).show();
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(Bitmap resource, Object model, Target<Bitmap> target, DataSource dataSource, boolean isFirstResource) {
                            return false;
                        }
                    })
                    .into(croppedImagePreview);
        } else {
            croppedImagePreview.setImageResource(R.drawable.add_image_illustration);
        }
    }

    public void showMessageEmptyTitle() {
        Toast.makeText(this, "please define a title", Toast.LENGTH_SHORT).show();
    }

    /*
     *********************************************************************************************
     * TEXTWATCHER
     *********************************************************************************************/
    private final TextWatcher titleWatcher = new TextWatcher() {
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        public void onTextChanged(CharSequence s, int start, int before, int count) {
            mShotEditionViewModel.onTitleChanged(s.toString());
        }

        public void afterTextChanged(Editable s) {
        }
    };

    private final TextWatcher tagWtacher = new TextWatcher() {
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        public void onTextChanged(CharSequence s, int start, int before, int count) {
            //todo - must manage tag limit in ViewmODEL
            String tagString = s.toString();
            int commas = s.toString().replaceAll("[^,]","").length();
            Timber.d(commas+"");
            if (commas<12) {
                mShotEditionViewModel.onTagChanged(tagString);
            } else {

            }
        }

        public void afterTextChanged(Editable s) {
        }
    };

    private final TextWatcher descriptionWatcher = new TextWatcher() {
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        public void onTextChanged(CharSequence s, int start, int before, int count) {
            mShotEditionViewModel.onDescriptionChanged(s.toString());
        }

        public void afterTextChanged(Editable s) {
        }
    };

}

File: app/src/main/java/seigneur/gauvain/mycourt/ui/shotEdition/EditShotActivity.kt
package seigneur.gauvain.mycourt.ui.shotEdition

import android.Manifest
import android.app.Activity
import android.app.Application
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.media.MediaPlayer
import android.net.Uri
import com.google.android.material.textfield.TextInputEditText
import androidx.core.app.ActivityCompat
import android.os.Bundle
import com.google.android.material.bottomsheet.BottomSheetBehavior
import androidx.core.widget.NestedScrollView
import androidx.appcompat.widget.Toolbar
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.*
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView


import com.yalantis.ucrop.UCrop

import java.io.File
import java.util.ArrayList

import javax.inject.Inject

import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.Optional
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.google.android.material.snackbar.Snackbar
import dagger.android.AndroidInjection
import droidninja.filepicker.FilePickerBuilder
import droidninja.filepicker.models.sort.SortingTypes
import seigneur.gauvain.mycourt.R
import seigneur.gauvain.mycourt.data.model.Attachment
import seigneur.gauvain.mycourt.data.model.Draft
import seigneur.gauvain.mycourt.ui.base.BaseActivity
import seigneur.gauvain.mycourt.ui.shotEdition.attachmentList.AttachmentItemCallback
import seigneur.gauvain.mycourt.ui.shotEdition.attachmentList.AttachmentsAdapter
import seigneur.gauvain.mycourt.ui.shotEdition.attachmentList.UnScrollableLayoutManager
import seigneur.gauvain.mycourt.utils.image.ImagePicker
import seigneur.gauvain.mycourt.ui.widget.FourThreeImageView
import seigneur.gauvain.mycourt.ui.widget.FourThreeVideoView
import seigneur.gauvain.mycourt.utils.*
import timber.log.Timber

class EditShotActivity : BaseActivity() , AttachmentItemCallback {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    @Inject
    lateinit var mApplication: Application

    private val mShotEditionViewModel: ShotEditionViewModel by lazy {
        ViewModelProviders.of(this, viewModelFactory).get(ShotEditionViewModel::class.java)
    }

    @BindView(R.id.toolbar)
    lateinit var mToolbar: Toolbar

    @BindView(R.id.main)
    lateinit var layoutMain: CoordinatorLayout

    @BindView(R.id.shot_title_edt)
    lateinit var mShotTitleEditor: TextInputEditText

    @BindView(R.id.shot_tag_edt)
    lateinit var mTagEditor: TextInputEditText

    @BindView(R.id.shot_description_edt)
    lateinit var mShotDescriptionEditor: TextInputEditText

    @BindView(R.id.cropped_img_preview)
    lateinit var croppedImagePreview: FourThreeImageView

    @BindView(R.id.video_view)
    lateinit var mVideoView: FourThreeVideoView

    @BindView(R.id.edit_shot_container)
    lateinit var mEditionContainer: NestedScrollView

    @BindView(R.id.bs_publish)
    lateinit var mBSPublish: LinearLayout

    @BindView(R.id.btn_store)
    lateinit var storeBtn: Button

    @BindView(R.id.btn_publish)
    lateinit var publishBtn : Button

    private lateinit var mBottomSheetBehaviour : BottomSheetBehavior<View>

    //Attachments
    @BindView(R.id.rv_attachment)
    lateinit var mRvAttachments: RecyclerView
    lateinit var mGridLayoutManager: GridLayoutManager
    var attachments=ArrayList<Attachment>()
    private val mAttachmentsAdapter: AttachmentsAdapter by lazy {
        AttachmentsAdapter(attachments, this)
    }

    /*
    *********************************************************************************************
    * LIFECYCLE
    *********************************************************************************************/
     override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //provide Dependencies
        AndroidInjection.inject(this)
        //Init ViewModel before inflate views to perform first request
        mShotEditionViewModel.init()
        setContentView(R.layout.activity_edit_shot)
        ButterKnife.bind(this)
        //Subscribe to ViewModel data and event only when View is inflated
        subscribeToLiveData(mShotEditionViewModel)
        subscribeToSingleEvent(mShotEditionViewModel)
        //set up listeners and behaviors
        setUpEditorListener()
        setUpBottomSheet()
        setUpScrollListener()
        initAttachmentList()
    }

    /**
     * Callback received when a permissions request has been completed.
     * @param requestCode
     * @param permissions
     * @param grantResults
     */
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            Constants.REQUEST_STORAGE_WRITE_ACCESS_PERMISSION -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                mShotEditionViewModel.onPermGranted()
            } else {
                mShotEditionViewModel.requestPerm()
            }
            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            when (requestCode) {
                Constants.PICK_IMAGE_REQUEST ->
                    ImagePicker.onImagePicked(data, mShotEditionViewModel)
                UCrop.REQUEST_CROP ->
                    ImagePicker.onImageCropped(data, mShotEditionViewModel)
                Constants.PICK_ATTACHMENT_REQUEST ->
                    ImagePicker.onAttachmentPicked(data, mShotEditionViewModel)
                else -> Timber.d("SELECTED MEDIA is null")
            }
        }
    }

    public override fun onDestroy() {
        super.onDestroy()
    }

    @OnClick(R.id.btn_store)
    fun store() {
        mShotEditionViewModel.onStoreDraftClicked()
    }

    @Optional
    @OnClick(R.id.btn_publish)
    fun publish() {
        mShotEditionViewModel.onPublishClicked()
    }

    @Optional
    @OnClick(R.id.cropped_img_preview)
    fun shotPreviewClick() {
        mShotEditionViewModel.onImagePreviewClicked()
    }


    /*
    *********************************************************************************************
    * EVENT WHICH VIEW WILL SUBSCRIBE
    *********************************************************************************************/
    private fun subscribeToLiveData(viewModel: ShotEditionViewModel) {
        viewModel.getUserType.observe(this,  Observer<Int> {
            manageEditionOption(it)
        })

        viewModel.getFileToUploadUri().observe(this, Observer<Uri> {
            if (FileUtils.getMimeType(it.toString()).equals(Constants.MP4)) {
                //todo if is video
                mVideoView.setVideoPath(it.toString())
                mVideoView.start()
                mVideoView.setOnErrorListener(mOnErrorListener)
                mVideoView.setOnPreparedListener{
                    //callback - video is ready to be played
                    Timber.d("is prepared")
                    it.isLooping =true //allow video to repeat
                }
            } else {
                Timber.d("image uri $it")
                this.displayShotImagePreview(it.toString())
            }
        })

        viewModel.title.observe(this, Observer<String> { Timber.d("title change:$it") })

        viewModel.description.observe(this,  Observer<String> { Timber.d("desc change:$it") })

        viewModel.tags.observe(this,  Observer<ArrayList<String>> { Timber.d("tags change:$it") })

        viewModel.attachmentsTobeUploaded.observe(this,  Observer<ArrayList<Attachment>> {
            updateAttachmentList(it)
        })

        viewModel.isReadyToPublish.observe(this,  Observer<Boolean?> {
            activePublishBottomSheet(it)
        })
    }

    private fun subscribeToSingleEvent(viewModel: ShotEditionViewModel) {
        viewModel.setUpUiCmd.observe(
                this, Observer { it ->
            manageUiFromDraftInfo(it)
        })

        viewModel.mPickShotCmd.observe(this, Observer {
            FilePickerBuilder.instance
                    .setMaxCount(1)
                    .setActivityTheme(R.style.LibAppTheme)
                    .showGifs(true)
                    //.enableVideoPicker(true) //API doesn't allow to upload VIDEO YET
                    .showFolderView(true)
                    .pickPhoto(this, Constants.PICK_IMAGE_REQUEST)
        })

        viewModel.mPickAttachmentCmd.observe(this, Observer {
            openAttachmentPicker()
        })

        viewModel.mCropImgCmd.observe(this,
                 Observer {
                    goToUCropActivity(
                            mShotEditionViewModel.mPickedFileMymeType,
                            FileUtils.getContentUriFromFilePath(this,
                                    mShotEditionViewModel.mPickedFileUri.toString()),
                            Uri.fromFile( File(this.cacheDir, mShotEditionViewModel.mPickedFileName)),
                            mShotEditionViewModel.mPickedImageDimens)
                })

        viewModel.pickCropImgErrorCmd.observe(this,
                Observer { Toast.makeText(this, getString(R.string.error_picking), Toast.LENGTH_SHORT).show() })

        viewModel.mCheckPerm.observe(this,
                Observer { checkPermissionExtStorage() })

        viewModel.mRequestPermCmd.observe(this,
                Observer {  requestPermission() })

        viewModel.onPublishSucceed.observe(this,
                Observer {
                    Toast.makeText(mApplication, "Publish suceed", Toast.LENGTH_SHORT)
                    finishAfterTransition()
                })

        viewModel.notifyUserNotReadyCmd.observe(this, Observer {
            val snack = Snackbar.make(findViewById(R.id.main),
                "Oops, please set a title and image", Snackbar.LENGTH_SHORT)
            SnackbarHelper.configSnackbar(this, snack)
            snack.show()

        })
    }

    /*
    *********************************************************************************************
    * AttachmentItemCallback
    *********************************************************************************************/
    override fun onAddClicked() {
        mShotEditionViewModel.onAddAttachmentClicked()
}

    override fun onAttachmentClicked(position: Int) {
    }

    override fun onAttachmentDeleted(position: Int) {
        mShotEditionViewModel.onRemoveAttachment(position)
    }

    /*
    *********************************************************************************************
    * REQUEST PERMISSION TO READ AND WRITE ON EXTERNAL STORAGE
    *********************************************************************************************/
    private fun checkPermissionExtStorage() {
        if (ActivityCompat.checkSelfPermission(this,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            mShotEditionViewModel.requestPerm()
        } else {
            mShotEditionViewModel.onPermGranted()
        }
    }

    private fun requestPermission() {
        requestPermission(
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                getString(R.string.permission_write_storage_rationale),
                Constants.REQUEST_STORAGE_WRITE_ACCESS_PERMISSION)
    }

    /*
    *********************************************************************************************
    * PICKER RELATED METHODS
    *********************************************************************************************/
    private fun openAttachmentPicker() {
        //Manage multiple selection later
        FilePickerBuilder.instance.setMaxCount(1)
                .setActivityTheme(R.style.LibAppTheme)
                .pickPhoto(this, Constants.PICK_ATTACHMENT_REQUEST)
    }

    private fun goToUCropActivity(
            imagePickedformat: String?,
            source: Uri?,
            destination: Uri,
            imageSize: IntArray?) {
        ImagePicker.goToUCropActivity(
                imagePickedformat,
                source!!,
                destination,
                this,
                imageSize!!)
    }

    private fun showImageNotUpdatable() {
        Toast.makeText(this@EditShotActivity, "Shot can't be changed in edition mode", Toast.LENGTH_SHORT).show()
    }


    private val mOnErrorListener = object : MediaPlayer.OnErrorListener {
        override fun onError(mp: MediaPlayer, what: Int, extra: Int): Boolean {
            // Your code goes here
            Timber.d("error loading video")
            return true
        }
    }

    /*
    *********************************************************************************************
    * EDITION UI RELATED FUNCTIONS
    *********************************************************************************************/
    /**
     * Manage editions options visibility in accordance to user type
     */
    private fun manageEditionOption(userType:Int) {
        when (userType) {
            Constants.USER_UPLOADER -> {
                mEditionContainer.visibility=View.VISIBLE
                mRvAttachments.visibility =View.GONE
                mBSPublish.visibility=View.VISIBLE
            }
            Constants.USER_PRO -> {
                mEditionContainer.visibility=View.VISIBLE
                mRvAttachments.visibility =View.VISIBLE
                mBSPublish.visibility=View.VISIBLE
            }
            else -> {
                mEditionContainer.visibility=View.GONE
                mBSPublish.visibility=View.GONE
            }
        }

    }

    /**
     * Manage UI in accordance to draft information defined in Draft
     */
    private fun manageUiFromDraftInfo(draft: Draft) {
        //set toolbar title
        if (draft.typeOfDraft == Constants.EDIT_MODE_NEW_SHOT) {
            mToolbar.title = getString(R.string.title_create_shot)
        } else {
            mToolbar.title = getString(R.string.title_edit)
        }
        //get title from draft object
        mShotTitleEditor.setText(draft.shot.title)
        //get description from draft objects
        val description = EditUtils.getDescription(draft)
        if (description != null && !description.isEmpty())
            mShotDescriptionEditor.setText(MyTextUtils.noTrailingwhiteLines(description))
        //get tags from draft objects
        mTagEditor.setText(EditUtils.getTagList(draft))
        //DISPLAY DEFAULT SHOT IMAGE FOR PUBLISHED SHOT AND NEW DRAFT
        //FOR DRAFT WITH STORED CROPPED IMAGE, WE GET IT FROM LIVEDATA
        if (draft.typeOfDraft == Constants.EDIT_MODE_UPDATE_SHOT) {
            displayShotImagePreview(draft.shot.imageHidpi) //high quality image displayed
        } else {
            if (draft.imageUri.isNullOrEmpty())
                displayShotImagePreview(null)
        }
    }

    private fun displayShotImagePreview(uriImageCropped: String?) {
       EditUtils.displayImage(uriImageCropped, this, croppedImagePreview)
    }

    private fun showMessageEmptyTitle() {
        Toast.makeText(this, "please define a title", Toast.LENGTH_SHORT).show()
    }

    private fun setUpBottomSheet() {
        mBottomSheetBehaviour = BottomSheetBehavior.from(mBSPublish)
        mBottomSheetBehaviour.state= BottomSheetBehavior.STATE_HIDDEN
        mBottomSheetBehaviour.setBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
            override fun onStateChanged(bottomSheet: View, newState: Int) {
            }
            override fun onSlide(bottomSheet: View, slideOffset: Float) {
            }
        })
    }

    private fun setUpScrollListener() {
        mEditionContainer.setOnScrollChangeListener(NestedScrollView.OnScrollChangeListener {
            v, scrollX, scrollY, oldScrollX, oldScrollY ->
            if (scrollY > oldScrollY) {
                Timber.i("Scroll DOWN")
                mBottomSheetBehaviour.state = BottomSheetBehavior.STATE_COLLAPSED
            }
            if (scrollY < oldScrollY) {
                Timber.i( "Scroll UP")
                mBottomSheetBehaviour.state = BottomSheetBehavior.STATE_COLLAPSED
            }

            if (scrollY == 0) {
                Timber.i("TOP SCROLL")
                mBottomSheetBehaviour.state = BottomSheetBehavior.STATE_HIDDEN

            }

            if (scrollY == v.getChildAt(0).measuredHeight - v.measuredHeight) {
                Timber.i("BOTTOM SCROLL")
                mBottomSheetBehaviour.state = BottomSheetBehavior.STATE_COLLAPSED
            }
        })
    }

    private fun activePublishBottomSheet(activate:Boolean?) {}

    /*
    *********************************************************************************************
    * TEXT WATCHER AND EDITION LISTENER
    *********************************************************************************************/
    private fun setUpEditorListener() {
        //Listen EditText
        mShotTitleEditor.addTextChangedListener(getTextWatcher(mShotTitleEditor))
        mTagEditor.addTextChangedListener(getTextWatcher(mTagEditor))
        mShotDescriptionEditor.addTextChangedListener(getTextWatcher(mShotDescriptionEditor))
    }

    private fun getTextWatcher(editText: TextInputEditText): TextWatcher {
        return object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                when(editText) {
                    mShotTitleEditor ->  mShotEditionViewModel.onTitleChanged(charSequence.toString())
                    mTagEditor ->  manageTagEdition(charSequence)
                    mShotDescriptionEditor -> mShotEditionViewModel.onDescriptionChanged(charSequence.toString())
                    else -> Timber.d("no action")
                }

            }
            override fun afterTextChanged(editable: Editable) {}
        }
    }

    private fun manageTagEdition(s: CharSequence) {
        val tagString = s.toString()
        val commas = s.toString().replace("[^,]".toRegex(), "").length
        Timber.d(commas.toString() + "")
        if (commas < 12) {
            mShotEditionViewModel.onTagChanged(tagString)
        } else {

        }
    }

    private fun initAttachmentList() {
        if (mRvAttachments.layoutManager==null && mRvAttachments.adapter==null) {
            mGridLayoutManager = UnScrollableLayoutManager(this, 5)
            mRvAttachments.layoutManager =  mGridLayoutManager
            mRvAttachments.adapter = mAttachmentsAdapter
            (mRvAttachments.layoutManager as UnScrollableLayoutManager).disableScrolling() //disable scroll
        }
    }

    private fun updateAttachmentList(newAttachmentList: ArrayList<Attachment>) {
        Timber.d("updateAttachmentList called")
        attachments.clear() //clear list
        for (i in newAttachmentList.indices) {
            //repopulate list
            attachments.add(newAttachmentList[i])
        }
        mAttachmentsAdapter.updateRv()
    }

}
